### ISSUE TITLE
Any interest in ECDSA?

### ISSUE THREAD

**Issue by jonasfj** [hueniverse/hawk#129]
I've been building a larger collection of services using hawk to do authentication, with a single auth-service holding all credentials where other services fetch the credentials from.  But I've found that as I want to let other services, less trusted, services use my auth-service (to manage credentials); that I would wish hawk had support for asymmetric algorithms.

ECDSA has high security with small 256 bit keys, and seems to be pretty fast; here is a demo I found using javascript:
http://kjur.github.io/jsrsasign/sample-ecdsa.html

So far I've used hawk and been very happy with it, because the key is relatively small 256 bit; which is easy to copy around, insert in environment variables, config files, etc.
But with 256bit ECDSA keys offering decent security, it's tempting to use that.

**TL;DR**
Would there be any interest in accepting a PR with support for ECDSA as `algorithm`?
(I fully understand if that is not the case)

**Comment by Kagami** [comment]
Hi. Sorry for the interruption. I see you guys here discussing ECC implementations in node (thanks for the links, btw).
I'm writting [simple isomorphic wrapper](https://github.com/bitchan/eccrypto) around `elliptic` (Browser) and `secp256k1-node` (Node) to work with K-256 curve. You may take a look.

At first I was trying to use native Node crypto module and WebCryptoAPI but:
* [It's hard to deal with ECDSA in Node](https://github.com/joyent/node/issues/6904): requires key to be in PEM format which is rather kludgy
* `secp256k1-node` [is faster](https://github.com/indutny/elliptic/pull/13) anyway (though `secp256k1` is still experimental)
* ECDH [is only supported in 0.11/0.12](https://github.com/joyent/node/pull/5854)
* WebCryptoAPI only defines P-\* curves for ECDHA/ECDH and it currently doesn't work in Firefox and Chrome anyway: [[1]](https://bugzilla.mozilla.org/show_bug.cgi?id=1034854), [[2]](https://code.google.com/p/chromium/issues/detail?id=338883)

**Comment by jonasfj** [comment]
ECDSA with secp256-k1 (as used by bitcoin) or EdDSA over Ed25519 as done by libsodium/NaCl, certainly seems like good cryptographic schemes for this purpose.

I started fooling around with hawk, trying to see if it would fit in nicely. But:
  * hawk is clearly architected for symmetric schemes,
  * features such as response signatures doesn't fit well with asymmetric schemes, and
  * considering the amount of changes involved it's unrealistic to expect other hawk implementations to catch up.

So I'm increasingly leaning towards not implementing this in hawk.

----
**Slightly off-topic**, but an alternative remedy to the issue with a single entity storing credentials and distributing them to multiple services can be solved with symmetric algorithms, by repeated application of HMAC the way AWS does it in their signature format version 4.

Basically, when authentication against example.com, you give hawk the key `HMAC(secret, 'example.com')`, that way the service running at example.com doesn't need to know `secret`. And the server storing credentials (hence, knows `secret`) just shares the value `HMAC(secret, 'example.com')` with example.com.

--- 
Anyways, considering that it would be hard to hawk implementions in python, ruby, php, etc. to implement an asymmetric scheme maybe we shouldn't try to stuff this into hawk.
